app/scripts/lib/rpc-method-middleware/handlers/wallet-requestPermissions.test.ts

```diff
@@ -0,0 +1,756 @@
+    const { handler, requestPermissionsForOrigin } = createMockedHandler();
+
+    await handler({
+      ...getBaseRequest(),
+      params: [
+        {
+          [Caip25EndowmentPermissionName]: {},
+        },
+      ],
+    });
+    expect(requestPermissionsForOrigin).toHaveBeenCalledWith({});
+  });
+
+  it('requests empty permissions from the PermissionController when no permissions are provided in params', async () => {
+    const { handler, requestPermissionsForOrigin } = createMockedHandler();
+
+    await handler({
+      ...getBaseRequest(),
+      params: [{}],
+    });
+    expect(requestPermissionsForOrigin).toHaveBeenCalledWith({});
+  });
+
+  it('does not update or grant a CAIP-25 endowment permission if eth_accounts and permittedChains approvals were not requested', async () => {
+    const { handler, updateCaveat, grantPermissions, getPermissionsForOrigin } =
+      createMockedHandler();
+
+    await handler({
+      ...getBaseRequest(),
+      params: [
+        {
+          otherPermission: {},
+        },
+      ],
+    });
+    expect(getPermissionsForOrigin).not.toHaveBeenCalled();
+    expect(updateCaveat).not.toHaveBeenCalled();
+    expect(grantPermissions).not.toHaveBeenCalled();
+  });
+
+  it('returns the granted permissions if eth_accounts and permittedChains approvals were not requested', async () => {
+    const { handler, response } = createMockedHandler();
+
+    await handler({
+      ...getBaseRequest(),
+      params: [
+        {
+          otherPermission: {},
+        },
+      ],
+    });
+    expect(response.result).toStrictEqual([
+      {
+        caveats: [{ value: { foo: 'bar' } }],
+        id: '2',
+        parentCapability: 'otherPermission',
+      },
+    ]);
+  });
+
+  it('does not update or grant a CAIP-25 endowment type permission if eth_accounts and permittedChains approvals were denied', async () => {
+    const {
+      handler,
+      updateCaveat,
+      grantPermissions,
+      getPermissionsForOrigin,
+      requestPermissionApprovalForOrigin,
+    } = createMockedHandler();
+    requestPermissionApprovalForOrigin.mockRejectedValue(
+      new Error('user denied approval'),
+    );
+
+    try {
+      await handler({
+        ...getBaseRequest(),
+        params: [
+          {
+            [RestrictedMethods.eth_accounts]: {},
+          },
+        ],
+      });
+    } catch (err) {
+      // noop
+    }
+    expect(getPermissionsForOrigin).not.toHaveBeenCalled();
+    expect(updateCaveat).not.toHaveBeenCalled();
+    expect(grantPermissions).not.toHaveBeenCalled();
+  });
+
+  describe('eth_accounts and permittedChains approvals were accepted', () => {
+    it('sets the approved chainIds on an empty CAIP-25 caveat with isMultichainOrigin: false if origin is not snapId', async () => {
+      const { handler } = createMockedHandler();
+
+      await handler(getBaseRequest());
```

Comment:
Similar as above, thoughts on making the `origin` explicit?

Reply:
https://github.com/MetaMask/metamask-extension/pull/27847/commits/f7a09fda54e6e4998b098b2ea84e8557b0e13d3d